assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml

```diff
@@ -81,6 +109,67 @@ spec:
         - name: ssl-certs-host
           mountPath: /etc/ssl/certs
           readOnly: true
+        - name: data
+          mountPath: /run/kube-apiserver
+      - name: haproxy
```

Comment:
I've read the full commit message (thank you very much for the big message explaining, I **really** appreciate it), I tried to follow the discussions linked there. Let me know if I'm missing something.

Also, thank you very much for trying so hard to fix this case. But now that the config to do it is here, I can't help to wonder... is it worth it? (Without this config, we wouldn't know how tricky it was or not, so thanks again for doing it! :))

The HAProxy as a sidecar only to manage avoid manual intervention when running a single controller and the update failed seems like a big cost to pay. We are expanding the attack surface (now HA Proxy bugs are very relevant to us), we need to care about keeping the version up to date and the helm whole chart config is more complicated.

The configuration is quite more complicated, actually (the 7443 port, the random IP to bind tricks, this not being a DaemonSet just to run multiple instances for the single node case).

Why do we care so much about upgrades to single controller nodes clusters that failed? Am I missing something?

Reply:
>Why do we care so much about upgrades to single controller nodes clusters that failed? Am I missing something?

Well, one argument for caring about it is, that this is what user with smaller cluster may want. Also, this way we don't need to have 2 different implementation for upgrades, but we just have one generic solution.

>The configuration is quite more complicated, actually (the 7443 port, the random IP to bind tricks, this not being a DaemonSet just to run multiple instances for the single node case).

Well, yes, but with help of the upstream, we can have exactly the same solution, but much simpler, and this is what I'm mainly hoping for.

We can perhaps re-evaluate it again.

Reply:
>What do you mean with exactly the same solution?

I mean we would have the exact same setup, with side container etc.

>If the HAProxy is removed (with a conditional in helm) for HA clusters, this same procedure will work, right? Or am I missing something?

If we combine it with `requiredDuringScheduling` pod anti affinity, then I _think_ yes. But if we do that, we could also revert the changes to the port setup. We would also needs to test scale up and scale down scenarios, when transitioning from single controller node to HA and back.

Reply:
> > If the HAProxy is removed (with a conditional in helm) for HA clusters, this same procedure will work, right? Or am I missing something?
> 
> If we combine it with `requiredDuringScheduling` pod anti affinity, then I _think_ yes. But if 

Are you trying to imply that if we don't use that, then no?

Looking more at it, it is using `hostNetwork: true` (see [here](https://github.com/kinvolk/lokomotive/pull/32/files#diff-12be5db389a0fefe0038d10211a890b4R51)), so the behaviour should not change.

So, if we remove this sidecar that is totally unneeded for tha HA setup (I'd argue that HA setup being stable and secure is more important than single node controller) it should work and there is no need for them to be used in HA setups. Right?

HA setup is way more important than single node setup, IMHO. Therefore, IMHO, I wouldn't let a single node temporal hack leak into HA setup and add a bigger attack surface and pain points (like upgrading just to upgrade HAProxy, etc.)

> we do that, we could also revert the changes to the port setup. We would also needs to test scale up and scale down scenarios, when transitioning from single controller node to HA and back.

What? This is totally unrelated, we never supported going back and forth from single node to HA and is not supported now, right? If this is correct, this seems totally to decide on this.

If we want to add support for that, we can see the trade-off of having it then (and it might be as simple as removing the `if` to get the sidecar on all cases, if we *really* want that. I'd say may not even in that case want that, and just a simple migration maybe with a *temporal* HAProxy, and then remove it. Having a more complex architecture than needed, for your most important case, to support other not so important cases doesn't seem like a good trade-off in general, IMHO).

Reply:
I wonder how `requiredDuringScheduling` behaves with helm upgrades BTW. If you request 3 replicas and one controller node will fail to bootstrap, it may hang and roll back the upgrades then. We should probably test that.

Regarding the HAProxy for HA setup, I don't mind removing it. What do you think @iaguis ?

Reply:
I agree with removing it if it works without it.

Reply:
_sticks "if/else" everywhere_